File: CheckIn.Api.Common/Models/Update/TaskUpdateModel.cs
using CheckIn.Api.Common.Enums;
using CheckIn.Api.Common.Models.Create;
using CheckIn.Api.Common.Models.Interfaces;

namespace CheckIn.Api.Common.Models.Update;

public record TaskUpdateModel : IEntityModel
{
    public Guid Id { get; init; }
    public required string Title { get; init; }
    public string? Notes { get; init; }
    public string? AllowedDomain { get; set; }
    public DateTime DeadLine { get; init; }
    public TaskState? State { get; init; }

    public bool RequiresAuthenticationToComplete { get; init; } = true;
    public List<string> InvitedEmails { get; init; } = new();
    public List<SubtaskTemplateCreateModel> Subtasks { get; init; } = new();
}

File: checkin_frontend/lib/core/models/task/task_update_model.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../enums/task_enums.dart';
import '../subtask_template/subtask_template_create_model.dart';

part 'task_update_model.freezed.dart';
part 'task_update_model.g.dart';

@freezed
sealed class TaskUpdateModel with _$TaskUpdateModel {
  const factory TaskUpdateModel({
    required String id,
    required String title,
    String? notes,
    required DateTime deadLine,
    @Default(true) bool requiresAuthenticationToComplete,
    TaskState? state,

    @Default([]) List<String> invitedEmails,
    @Default([]) List<SubtaskTemplateCreateModel> subtasks,
    String? allowedDomain,

    @JsonKey(includeToJson: false) @Default(null) bool? isDomainValid,
  }) = _TaskUpdateModel;

  factory TaskUpdateModel.fromJson(Map<String, dynamic> json) =>
      _$TaskUpdateModelFromJson(json);
}